src/visit.rs

```diff
@@ -435,6 +435,29 @@ impl<'ast> Visit<'ast> for DiscoveryVisitor<'_> {
             .for_each(|rep| self.collect_mutant(i.op.span().into(), rep, Genre::BinaryOperator));
         syn::visit::visit_expr_binary(self, i);
     }
+
+    fn visit_expr_unary(&mut self, i: &'ast syn::ExprUnary) {
+        let _span = trace_span!("unary", line = i.op.span().start().line).entered();
+        trace!("visit unary operator");
+        if attrs_excluded(&i.attrs) {
+            return;
+        }
+        let replacements = match i.op {
+            UnOp::Not(_) => vec![quote! {}, quote! {-}],
```

Comment:
I was originally thinking that we would only generate deletions of `!` and `-`, not replace one with the other. It's an interesting idea, and it would be potentially viable for signed integer types, that can be either inverted or negated. 

I am a little concerned that this might generate a lot of unviable mutants because, I would guess, `!` is much more often applied to booleans, and negating them is not legal. Probably this is caught fairly early and relatively cheap, but, still, if it's mostly a waste then let's not. Also, I would think that code that's supposed to do a bitwise inversion of an integer would be adequately exercised by just not inverting...

Reply:
You make a good point that in cases where swapping `!` with `-` (in either direction) is valid, you would expect simply removing the operator to be a sufficient mutation.

> When adding a new mutation pattern like this I think it's worth running the code over some significant tree/s to see if the generated mutants generally seem like they might find real bugs and are reasonably often viable.

When you say significant trees do you mean external projects? I did of course look at the mutations generated for the trees in `testdata`, but I can also see what it looks like in some other projects.

Reply:
Yes, some external trees where you feel you know something about them, or the cargo mutants tree itself, or some popular rust crates like itertools.

Reply:
Ok I did a bit of testing of the `!` <=> `-` mutation
- [`itertools`](https://github.com/rust-itertools/itertools): 0/28 viable
- [`rand`](https://github.com/rust-random/rand): 0/7 viable
- [`hashbrown`](https://github.com/rust-lang/hashbrown): 3/11 viable (all 'missed')
  - https://github.com/rust-lang/hashbrown/blob/97c21406c6bc5945d000373f324f7ae5cd25b60f/src/external_trait_impls/rayon/set.rs#L257
  - https://github.com/rust-lang/hashbrown/blob/97c21406c6bc5945d000373f324f7ae5cd25b60f/src/external_trait_impls/rayon/set.rs#L107
  - https://github.com/rust-lang/hashbrown/blob/97c21406c6bc5945d000373f324f7ae5cd25b60f/src/raw/mod.rs#L3799
- [`serde`](https://github.com/serde-rs/serde): 0/1 viable

Which is a pretty low rate of viability overall. Maybe crates that do more bitwise arithmetic (like hashing) could expect to see some viable cases (and indeed, `hashbrown` does).

This is making me feel like these mutations shouldn't be included. Let me know if you feel otherwise.